### PR TITLE
Add expand_dims utility

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

## Category:

**New feature**

## Description:

- Adds a new utility to expand the number of dimensions of a TensorShape/TensorListShape, by inserting dimensions with extent one to a specified location
- To be used in https://github.com/NVIDIA/DALI/pull/4492/

### Affected modules and functionalities:

No modules affected, standalone utility

### Key points relevant for the review:

All of it

### Tests:

dali/core/expand_dims_test.cc
- [ ] Existing tests apply
- [x] New tests added
  - [ ] Python tests
  - [x] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [x] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [x] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

File: dali/pipeline/data/view_as_higher_ndim.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
...
+#include <vector>
+#include "dali/pipeline/data/views.h"
+
+namespace dali {
+
+/**
+ * @brief Inserts `insert_ndim` new dimensions with extent one to a given position in a shape
+ */
+static TensorListShape<> insert_unit_dims(const TensorListShape<> &sh, int insert_ndim,
+                                          int position) {
```
Comment (mzient) [inline dali/pipeline/data/view_as_higher_ndim.h:28]:
```suggestion
template <int out_ndim = DynamicDimensions, int in_ndim>
inline TensorListShape<out_ndim> insert_unit_dims(const TensorListShape<in_ndim> &sh, int position, int count) {
```
...and move to `tensor_shape.h`

Comment (jantonguirao) [reply dali/pipeline/data/view_as_higher_ndim.h]: As discussed offline, renamed to `expand_dims` and moved to dali/core

**Review by stiepan — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build

File: include/dali/core/expand_dims.h [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+
+#ifndef DALI_CORE_EXPAND_DIMS_H_
+#define DALI_CORE_EXPAND_DIMS_H_
+
+#include "dali/core/tensor_shape.h"
+
+namespace dali {
+
+/**
+ * @brief  Inserts new dimensions with extent one to the new shape as needed to match a given number
```
Comment (mzient) [inline include/dali/core/expand_dims.h:23]: Nitpick:
```suggestion
 * @brief  Inserts new dimensions with extent 1 to the new shape as needed to match a given number
```
I think it looks a tiny bit less WTF this way.

File: include/dali/core/expand_dims.h [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+
+ *
+ * @param sh input shape
+ * @param new_dims_position Position where the new dimensions are inserted
+ * @param dynamic_out_dim Dynamic number of dimensions in the output (relevant if static_out_ndim ==
+ DynamicDimensions)
+ */
+template <int static_out_ndim = DynamicDimensions, int static_in_ndim = DynamicDimensions>
+TensorListShape<static_out_ndim> expand_dims(const TensorListShape<static_in_ndim> &sh,
+                                             int new_dims_position = 0, int dynamic_out_dim = -1) {
```
Comment (mzient) [inline include/dali/core/expand_dims.h:34]: To consider: add a variant that fills an output parameter and then implement this as:
```C++
TensorListShape<static_out_ndim> expand_dims(const TensorListShape<static_in_ndim> &sh,
                                             int new_dims_position = 0, int dynamic_out_dim = -1) {
  TensorListShape<static_out_ndim> new_sh;
  expand_dims(new_sh, sh, new_dims_position, dynamic_out_dim);
  return new_sh;
}
```

File: include/dali/core/expand_dims.h [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+      new_sample_sh[out_d] = 1;
+    }
+  }
+  return new_sh;
+}
+
+
+template <int static_out_ndim = DynamicDimensions, int static_in_ndim = DynamicDimensions>
+TensorShape<static_out_ndim> expand_dims(const TensorShape<static_in_ndim> &sh,
+                                         int new_dims_position = 0, int dynamic_out_dim = -1) {
```
Comment (mzient) [inline include/dali/core/expand_dims.h:74]: Likewise.

File: include/dali/core/expand_dims.h [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef DALI_CORE_EXPAND_DIMS_H_
+#define DALI_CORE_EXPAND_DIMS_H_
+
+#include "dali/core/tensor_shape.h"
```
Comment (mzient) [inline include/dali/core/expand_dims.h:18]:
```suggestion
#include <stdexcept>
#include "dali/core/format.h"
#include "dali/core/tensor_shape.h"
```

Comment (jantonguirao) [reply include/dali/core/expand_dims.h:20]: Done

**Comment by jantonguirao** [comment]
!build

**Pull request merged**